(PDB-2678) Initial checkin of the caching proxy

senior opened the pull request:
The code lazily stores crates from crates.io as they are requested. Storing them locally after retrieving them from the remote server. The cache directory and the port can be changed via environment variables and activity is logged to the logs directory using logback.

Commit:
(PDB-2678) Initial checkin of the caching proxy

The code lazily stores crates from crates.io as they are
requested. Storing them locally after retrieving them from the remote
server. The cache directory and the port can be changed via environment
variables and activity is logged to the logs directory using logback.

Commit:
(PDB-2678) Added cache hit/miss tests

Added tests that validate the redirect logic, persisting new crates and
retriving previously hashed crates.

ajroetker commented on project.clj:
```diff
@@ -0,0 +1,14 @@
+                 [ch.qos.logback/logback-classic "1.1.7"]
+                 [environ "1.0.3"]]
+  :profiles {:dev {:dependencies [[ring/ring-mock "0.3.0"]]}}
+  :aliases {"crate-proxy" ["trampoline" "run" "-m" "crates-caching-proxy.core"]})
```
I think you need a `puppetlabs.crates-ca...` here, the alias is broken.

senior replied:
Yeah, I renamed it to puppetlabs... to be similar to our other projects, forgot to update that part, I'll fix it up

ajroetker commented on README.md:
```diff
@@ -1 +1,21 @@
+that there are two pieces to downloading crates, the first is the
+[create-io.index](https://github.com/rust-lang/crates.io-index) and
+the second is the packages themselves. This project is only concerned
+with storing the crates themselves, not the index.
```
We might want instructions on how to setup the whole stack, end-to-end.

senior replied:
So you're thinking the manifest with the cron setup etc?

ajroetker replied:
Well I was more thinking of just directions of how one could set this up locally. Like moving the `config.json` into place.

ajroetker commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,99 @@
+
+(def cache-root (delay (get environ/env :cache-root "proxy-cache")))
+
+(def port (delay (Long/parseLong (get environ/env :crate-proxy-port "8080"))))
```
Could we make this 8888 or something, just so it doesn't collide with PDB by default :)

senior replied:
sure

Commit:
(PDB-2678) Switched default port to 8888

Commit:
(PDB-2678) Updated README with cargo instructions

The README included directions on starting this service, but didn't have
anything on how to setup cargo to use this proxy. That information with
some examples is now included in the README.

ajroetker commented on README.md:
````diff
@@ -1 +1,52 @@
+
+```
+[registry]
+index = "file:///local/path/to/creates.io-index
````
missing a " at the end of this line :)

ajroetker commented on README.md:
````diff
@@ -1 +1,52 @@
+```
+[registry]
+index = "file:///local/path/to/creates.io-index
+```"
````
I think the double quote ended up down here

ajroetker commented:
This works for me as expected. I'm :+1: module @rbrw's comment about adding a line wrt the file renaming this morning.

rbrw commented on resources/logback.xml:
```diff
@@ -0,0 +1,24 @@
+<configuration scan="true">
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
```
I forget, does this actually go to stdout?  If so, by default, I might normally send it to stderr if it's easy.

senior replied:
It goes to System.out by default (the docs indicate that [here](http://logback.qos.ch/manual/appenders.html#ConsoleAppender)). This file is just a copy of the logback file we include in all of our projects, best I can tell, all of them use the default to STDOUT (examples from PDB: [1](https://github.com/puppetlabs/puppetdb/blob/master/resources/logback.xml#L2),[2](https://github.com/puppetlabs/puppetdb/blob/master/resources/ext/config/logback.xml),[3](https://github.com/puppetlabs/puppetdb/blob/master/test-resources/logback-test.xml)). I'm not using the STDOUT appender, though it matches our copy/pasted logback config. Line 22 below only references the file appender. My guess is why that was there originally is for easier developer debugging using the log, not sure.  I'm not opposed to adding the System.err stuff, but it seems like we should do that separately if there's value in it.

rbrw replied:
OK, well in general, I'd send bits like logging to stderr, reserving stdout for possible future material information, i.e. `foo=$(wc -l < something)`.  If logging goes to stderr by default, no command line arguments can ever produce (easily) useful information like that.

rbrw commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,99 @@
+            [slingshot.slingshot :refer [throw+ try+]]
+            [environ.core :as environ]))
+
+(def cache-root (delay (get environ/env :cache-root "proxy-cache")))
```
Might be some broader issue I'm missing, but I'd have assumed that System/getenv is very cheap, at least on Linux -- assuming it's really just a veneer over normal environment variables (i.e. getenv()).

rbrw replied:
Also tempted to suggest a default destination of just "cache", which should be fine if you're running it from the source tree, and if you're not, you might want something else entirely, say "crates", "crates-cache" or whatever.

senior replied:
I used the delay above more for AOT reasons, might be paranioa, do some testing and see if the scenario I'm worried about can happen.

senior replied:
cache sounds fine as a subdirectory to me, should probably include it in the gitignore as well

rbrw commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,99 @@
+        final-crate-path (path-get @cache-root package version package-name)]
+    (Files/createDirectories crate-path (into-array FileAttribute []))
+    (let [temp-file (create-temp-file crate-path package "crate")]
+      (with-open [input-stream (download-crate package version)]
```
Might wrap this and the move in a (try ... (finally (Files/deleteIfExists temp-file)), so we can't leave detritus on error.

senior replied:
Good catch, I'll add that

rbrw commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,99 @@
+(defn find-crate [package version package-name]
+  (if-let [file-response (create-file-response package version package-name)]
+    (do
+      (log/infof "Found version '%s' of '%s' cached, returning cached version" version package-name)
```
Wonder about just using (pr-str x) with %s to handle the quotes everywhere.  The normal case should look normal, and others will be unambiguous.

rbrw commented:
Some minor comments, but overall this looks pretty good (I might also be inclined to wrap some of the longer lines).

Commit:
(PDB-2678) Change default cache dir to 'cache'

This commit changes the default directory from 'proxy-cache' to
'cache'. It also simplifies the retrieval of environment variables and
removes the dependency on environ. There was a possibility of errors
causing temp files to be left on disk. This commit adds a try/finally
block to ensure those temp files get cleaned up in the event of a failure

senior commented:
@rbrw fixed it up based on your comments, removed the environ dependency and simplified the environment variables interaction, wrapped the copy/move in a new function that ensures the temp files are deleted.

rbrw commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,121 @@
+               :response-body (slurp (:body resp))}
+              (format "Error downloading package '%s'" package-url)))))
+
+(defn ensure-removed [^Path path f]
```
Wondered about with-final-deletion or with-final-removal or whatever (providing auto-indent as a bonus).

rbrw commented on src/puppetlabs/crates_caching_proxy/core.clj:
```diff
@@ -0,0 +1,121 @@
+(defn -main []
+  (let [startup-log-str (format "Starting the caching proxy with '%s' as the proxy's cache directory, on port '%s'"
+                                cache-root port)]
+    (println startup-log-str)
```
Typically, I'd send this to `*err*` too (for the reasons above).
